src/CP15.cpp

```diff
@@ -332,153 +366,596 @@ u32 ARMv5::RandomLineIndex()
-        return;
-    }
-    if (ICacheTags[id+3] == tag)
-    {
-        CodeCycles = 1;
-        CurICacheLine = &ICache[(id+3) << 5];
-        return;
+        if ((ICacheTags[id+set] & ~(CACHE_FLAG_DIRTY_MASK | CACHE_FLAG_SET_MASK)) == (tag | CACHE_FLAG_VALID))
+        {
+            CodeCycles = 1;
+            u32 *cacheLine = (u32 *)&ICache[(id+set) << ICACHE_LINELENGTH_LOG2];
+            if (CP15BISTTestStateRegister & CP15_BIST_TR_DISABLE_ICACHE_STREAMING)
+            {
+                // Disabled ICACHE Streaming:
+                // retreive the data from memory, even if the data was cached
+                // See arm946e-s Rev 1 technical manual, 2.3.15 "Register 15, test State Register")
+                CodeCycles = NDS.ARM9MemTimings[tag >> 14][2]; 
+                if (CodeMem.Mem)
+                {
+                    return *(u32*)&CodeMem.Mem[(addr & CodeMem.Mask) & ~3];
+                } else
+                {
+                    return NDS.ARM9Read32(addr & ~3);
+                }     
+            }
+            return cacheLine[(addr & (ICACHE_LINELENGTH -1)) >> 2];
+        }
     }
 
     // cache miss
 
+    // We do not fill the cacheline if it is disabled in the 
+    // BIST test State register (See arm946e-s Rev 1 technical manual, 2.3.15 "Register 15, test State Register")
+    if (CP15BISTTestStateRegister & CP15_BIST_TR_DISABLE_ICACHE_LINEFILL)
+    {
+        CodeCycles = NDS.ARM9MemTimings[tag >> 14][2]; 
+        if (CodeMem.Mem)
+        {
+            return *(u32*)&CodeMem.Mem[(addr & CodeMem.Mask) & ~3];
+        } else
+        {
+            return NDS.ARM9Read32(addr & ~3);
+        }        
+    }
+
     u32 line;
-    if (CP15Control & (1<<14))
+#if 0
+    // caclulate in which cacheline the data is to be filled
+    // The code below is doing the same as the if-less below
+    // It increases performance by reducing banches. 
+    // The code is kept here for readability.
+    // 
+    // NOTE: If you need to update either part, you need
+    //       to update the other too to keep them in sync!
+    //       
+
+    if (CP15Control & CP15_CACHE_CR_ROUNDROBIN)
     {
-        line = ICacheCount[id>>2];
-        ICacheCount[id>>2] = (line+1) & 0x3;
+        line = ICacheCount;
+        ICacheCount = (line+1) & (ICACHE_SETS-1);
     }
     else
     {
         line = RandomLineIndex();
     }
 
-    line += id;
+    if (ICacheLockDown)
+    {
+        if (ICacheLockDown & CACHE_LOCKUP_L)
+        {
+            // load into locked up cache
+            // into the selected set
+            line = ICacheLockDown & (ICACHE_SETS-1);
+        } else
+        {
+            u8 minSet = ICacheLockDown & (ICACHE_SETS-1);
+            line = line | minSet;
+        }
+    }
+
+#else
+    // Do the same as above but instead of using if-else 
+    // utilize the && and || operators to skip parts of the operations
+    // With the order of comparison we can put the most likely path
+    // checked first
```

Comment:
I'm not really sure whether this abusing shortcircuit evaluation really gives better code. If you want to steer the compiler using likely and unlikely markers could be used for the ifs.

Reply:
I have removed this and added the [[(un)likely]] attributes